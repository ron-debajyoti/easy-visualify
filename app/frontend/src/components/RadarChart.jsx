import React, { useState } from 'react';
import styled from 'styled-components/macro';
import PropTypes from 'prop-types';
import { Radar } from 'react-chartjs-2';

const CheckBoxWrapper = styled.div`
  display: flex;
  align-items: center;
  position: relative;
  margin: 10px;
`;

const Wrapper = styled.div`
  float: right;
  margin-right: auto;
`;

const Message = styled.h5`
  display: flex;
  font-weight: 75;
  color: white;
  margin: 5px;
`;

const CheckBoxLabel = styled.label`
  float: right;
  width: 42px;
  height: 26px;
  border-radius: 15px;
  background: #bebebe;
  cursor: pointer;
  &::after {
    content: '';
    display: block;
    border-radius: 50%;
    width: 18px;
    height: 18px;
    margin: 3px;
    background: #ffffff;
    box-shadow: 1px 3px 3px 1px rgba(0, 0, 0, 0.2);
    transition: 0.2s;
  }
`;
const CheckBox = styled.input`
  display: flex;
  opacity: 0;
  z-index: 1;
  border-radius: 15px;
  width: 42px;
  height: 26px;
  &:checked + ${CheckBoxLabel} {
    background: #4fbe79;
    &::after {
      content: '';
      display: block;
      border-radius: 50%;
      width: 18px;
      height: 18px;
      margin-left: 21px;
      transition: 0.2s;
    }
  }
`;

const ChartTitle = styled.h3`
  text-align: center;
`;

/* PropTypes validation */
const propTypes = {
  username: PropTypes.string.isRequired,
  // eslint-disable-next-line react/forbid-prop-types
  chartData: PropTypes.object.isRequired,
};

const getUserChartData = (chartData, username) => {
  const { labels, datasets } = chartData;
  const userDataCharts = datasets.filter((dataset) => dataset.owner === username);
  return { labels, datasets: userDataCharts };
};

const RadarChart = (props) => {
  const [toggle, setToggle] = useState(false);
  const { chartData, username } = props;
  const userChartData = getUserChartData(chartData, username);

  const handleOnChange = () => setToggle(!toggle);

  const RenderChart = (click) => {
    const { type } = click;
    if (type) {
      return (
        <Radar
          data={userChartData}
          options={{
            legend: {
              scale: { pointLabels: { fontSize: 20 } },
              labels: {
                font: 20,
                fontSize: window.innerWidth > 350 ? 20 : 10,
              },
            },
          }}
        />
      );
    }
    return (
      <Radar
        data={chartData}
        options={{
          legend: {
            scale: { pointLabels: { fontSize: 20 } },
            labels: {
              font: 20,
              fontSize: window.innerWidth > 350 ? 20 : 10,
            },
          },
        }}
      />
    );
  };

  return (
    <div>
      <ChartTitle style={{ color: 'white' }}>
        {' '}
        Audio Analysis of Tracks and Saved Playlists{' '}
      </ChartTitle>
      <Wrapper>
        <Message>User created playlists</Message>
        <CheckBoxWrapper>
          <CheckBox id="checkbox" type="checkbox" onChange={() => handleOnChange()} />
          <CheckBoxLabel htmlFor="checkbox" />
        </CheckBoxWrapper>
      </Wrapper>
      <RenderChart type={toggle} />
    </div>
  );
};

RadarChart.propTypes = propTypes;

export default RadarChart;
