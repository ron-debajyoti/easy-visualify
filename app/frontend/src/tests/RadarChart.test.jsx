import React from 'react';
import renderer from 'react-test-renderer';
import sinon from 'sinon';
import { shallow } from 'enzyme';
import { unmountComponentAtNode } from 'react-dom';
import RadarChart from '../components/RadarChart';
import { fakeRadarChartData } from './fakedata';

let container = null;

beforeEach(() => {
  // setup a DOM element as a render target
  container = document.createElement('div');
  document.body.appendChild(container);
});

afterEach(() => {
  // cleanup on exiting
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});

it('renders the Radar Chart without crashing', () => {
  // snapshot contains toggle-false
  const tree = renderer
    .create(<RadarChart chartData={fakeRadarChartData} username="trialUsername" />, container)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders the chart correctly before/after toggling', () => {
  const onChange = sinon.spy();

  // act(() =>
  //   render(<RadarChart chartData={fakeRadarChartData} username="trialUsername" />, container)
  // );
  const wrapped = shallow(<RadarChart chartData={fakeRadarChartData} username="trialUsername" />, {
    lifecycleExperimental: true,
  });

  expect(wrapped.find('[className="checkbox"]')).toBeTruthy();
  wrapped.find('[className="checkbox"]').simulate('change', { target: { checked: true } });
  expect(wrapped.find('[className="toggle-true"]')).toBeTruthy();
  expect(onChange).toBeTruthy();
});
