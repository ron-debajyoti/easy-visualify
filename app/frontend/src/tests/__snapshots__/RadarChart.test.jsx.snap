// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the Radar Chart without crashing 1`] = `
.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: relative;
  margin: 10px;
}

.c1 {
  float: right;
  margin-right: auto;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-weight: 75;
  color: white;
  margin: 5px;
}

.c6 {
  float: right;
  width: 42px;
  height: 26px;
  border-radius: 15px;
  background: #bebebe;
  cursor: pointer;
}

.c6::after {
  content: '';
  display: block;
  border-radius: 50%;
  width: 18px;
  height: 18px;
  margin: 3px;
  background: #ffffff;
  box-shadow: 1px 3px 3px 1px rgba(0,0,0,0.2);
  -webkit-transition: 0.2s;
  transition: 0.2s;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  opacity: 0;
  z-index: 1;
  border-radius: 15px;
  width: 42px;
  height: 26px;
}

.c4:checked + .c5 {
  background: #4fbe79;
}

.c4:checked + .c5::after {
  content: '';
  display: block;
  border-radius: 50%;
  width: 18px;
  height: 18px;
  margin-left: 21px;
  -webkit-transition: 0.2s;
  transition: 0.2s;
}

.c0 {
  text-align: center;
}

@media (max-width:768px) {
  .c7 {
    min-height: 60vh;
    min-width: 40vh;
  }
}

@media (min-width:768px) {
  .c7 {
    min-height: 85vh;
  }
}

@media (max-width:768px) {
  .c2 {
    font-size: small;
  }
}

@media (min-width:768px) {
  .c2 {
    font-size: medium;
  }
}

<div
  className="radar-chart"
>
  <h3
    className="c0"
    style={
      Object {
        "color": "white",
      }
    }
  >
     
    Audio Analysis of Tracks and Saved Playlists
     
  </h3>
  <div
    className="c1"
  >
    <h5
      className="c2"
    >
      User created playlists
    </h5>
    <div
      className="c3"
    >
      <input
        className="c4 checkbox"
        id="checkbox"
        onChange={[Function]}
        type="checkbox"
      />
      <label
        className="c5 c6"
        htmlFor="checkbox"
      />
    </div>
  </div>
  <div
    className="c7 toggle-false"
  >
    <canvas
      data-testid="canvas"
      height={150}
      onClick={[Function]}
      role="img"
      width={300}
    />
  </div>
</div>
`;
